File: internal/server/dto/generator.go
package dto

import (
	"time"

	"github.com/topvennie/sortifyr/internal/database/model"
	"github.com/topvennie/sortifyr/pkg/utils"
)

type GeneratorWindow struct {
	Start             time.Time `json:"start"`
	End               time.Time `json:"end"`
	MinPlays          int       `json:"min_plays" validate:"min=0"`
	BurstIntervalDays int       `json:"burst_interval_days"`
}

func generatorWindowDTO(g model.GeneratorWindow) GeneratorWindow {
	return GeneratorWindow{
		Start:             g.Start,
		End:               g.End,
		MinPlays:          g.MinPlays,
		BurstIntervalDays: int(g.BurstInterval.Hours() / 24),
	}
}

func (g GeneratorWindow) ToModel() *model.GeneratorWindow {
	return &model.GeneratorWindow{
		Start:         g.Start,
		End:           g.End,
		MinPlays:      g.MinPlays,
		BurstInterval: time.Duration(g.BurstIntervalDays) * 24 * time.Hour,
	}
}

type GeneratorPresetCustomParams struct{}

func generatorPresetCustomParamsDTO(params *model.GeneratorPresetCustomParams) *GeneratorPresetCustomParams {
	if params == nil {
		return nil
	}

	return &GeneratorPresetCustomParams{}
}

func (g GeneratorPresetCustomParams) ToModel() *model.GeneratorPresetCustomParams {
	return &model.GeneratorPresetCustomParams{}
}

type GeneratorPresetForgottenParams struct{}

func generatorPresetForgottenParamsDTO(params *model.GeneratorPresetForgottenParams) *GeneratorPresetForgottenParams {
	if params == nil {
		return nil
	}

	return &GeneratorPresetForgottenParams{}
}

func (g GeneratorPresetForgottenParams) ToModel() *model.GeneratorPresetForgottenParams {
	return &model.GeneratorPresetForgottenParams{}
}

type GeneratorPresetTopParams struct {
	Window GeneratorWindow `json:"window"`
}

func generatorPresetTopParamsDTO(params *model.GeneratorPresetTopParams) *GeneratorPresetTopParams {
	if params == nil {
		return nil
	}

	return &GeneratorPresetTopParams{
		Window: generatorWindowDTO(params.Window),
	}
}

func (g GeneratorPresetTopParams) ToModel() *model.GeneratorPresetTopParams {
	return &model.GeneratorPresetTopParams{
		Window: *g.Window.ToModel(),
	}
}

type GeneratorPresetOldTopParams struct {
	PeakWindow   GeneratorWindow `json:"peak_window"`
	RecentWindow GeneratorWindow `json:"recent_window"`
}

func generatorPresetOldTopParamsDTO(params *model.GeneratorPresetOldTopParams) *GeneratorPresetOldTopParams {
	if params == nil {
		return nil
	}

	return &GeneratorPresetOldTopParams{
		PeakWindow:   generatorWindowDTO(params.PeakWindow),
		RecentWindow: generatorWindowDTO(params.RecentWindow),
	}
}

func (g GeneratorPresetOldTopParams) ToModel() *model.GeneratorPresetOldTopParams {
	return &model.GeneratorPresetOldTopParams{
		PeakWindow:   *g.PeakWindow.ToModel(),
		RecentWindow: *g.RecentWindow.ToModel(),
	}
}

type GeneratorParams struct {
	TrackAmount         int   `json:"track_amount" validate:"min=0"`
	ExcludedPlaylistIDs []int `json:"excluded_playlist_ids,omitzero"`
	ExcludedTrackIDs    []int `json:"excluded_track_ids,omitzero"`

	Preset model.GeneratorPreset `json:"preset" validate:"required"`

	ParamsCustom    *GeneratorPresetCustomParams    `json:"params_custom,omitzero"`
	ParamsForgotten *GeneratorPresetForgottenParams `json:"params_forgotten,omitzero"`
	ParamsTop       *GeneratorPresetTopParams       `json:"params_top,omitzero"`
	ParamsOldTop    *GeneratorPresetOldTopParams    `json:"params_old_top,omitzero"`
}

func generatorParamsDTO(params model.GeneratorParams) GeneratorParams {
	return GeneratorParams{
		TrackAmount:         params.TrackAmount,
		ExcludedPlaylistIDs: params.ExcludedPlaylistIDs,
		ExcludedTrackIDs:    params.ExcludedTrackIDs,
		Preset:              params.Preset,
		ParamsCustom:        generatorPresetCustomParamsDTO(params.ParamsCustom),
		ParamsForgotten:     generatorPresetForgottenParamsDTO(params.ParamsForgotten),
		ParamsTop:           generatorPresetTopParamsDTO(params.ParamsTop),
		ParamsOldTop:        generatorPresetOldTopParamsDTO(params.ParamsOldTop),
	}
}

func (g GeneratorParams) ToModel() model.GeneratorParams {
	var paramsCustom *model.GeneratorPresetCustomParams
	if g.ParamsCustom != nil {
		paramsCustom = g.ParamsCustom.ToModel()
	}
	var paramsForgotten *model.GeneratorPresetForgottenParams
	if g.ParamsForgotten != nil {
		paramsForgotten = g.ParamsForgotten.ToModel()
	}
	var paramsTop *model.GeneratorPresetTopParams
	if g.ParamsTop != nil {
		paramsTop = g.ParamsTop.ToModel()
	}
	var paramsOldTop *model.GeneratorPresetOldTopParams
	if g.ParamsOldTop != nil {
		paramsOldTop = g.ParamsOldTop.ToModel()
	}
	return model.GeneratorParams{
		TrackAmount:         g.TrackAmount,
		ExcludedPlaylistIDs: g.ExcludedPlaylistIDs,
		ExcludedTrackIDs:    g.ExcludedTrackIDs,
		Preset:              g.Preset,
		ParamsCustom:        paramsCustom,
		ParamsForgotten:     paramsForgotten,
		ParamsTop:           paramsTop,
		ParamsOldTop:        paramsOldTop,
	}
}

type Generator struct {
	ID              int             `json:"id"`
	Name            string          `json:"name" validate:"required"`
	Description     string          `json:"description,omitzero"`
	PlaylistID      int             `json:"playlist_id,omitzero"`
	IntervalDays    int             `json:"interval_days"`
	SpotifyOutdated bool            `json:"spotify_outdated"`
	Params          GeneratorParams `json:"params" validate:"required"`
	Tracks          []Track         `json:"tracks"`
	UpdatedAt       time.Time       `json:"updated_at"`
}

func GeneratorDTO(gen *model.Generator) Generator {
	return Generator{
		ID:              gen.ID,
		Name:            gen.Name,
		Description:     gen.Description,
		PlaylistID:      gen.PlaylistID,
		IntervalDays:    int(gen.Interval.Hours() / 24),
		SpotifyOutdated: gen.SpotifyOutdated,
		Params:          generatorParamsDTO(gen.Params),
		Tracks:          utils.SliceMap(gen.Tracks, func(t model.Track) Track { return TrackDTO(&t) }),
		UpdatedAt:       gen.UpdatedAt,
	}
}

type GeneratorSave struct {
	ID             int             `json:"id"`
	Name           string          `json:"name" validate:"required"`
	Description    string          `json:"description"`
	CreatePlaylist bool            `json:"create_playlist"`
	IntervalDays   int             `json:"interval_days" validate:"min=0"`
	Params         GeneratorParams `json:"params" validate:"required"`
}

func (g GeneratorSave) ToModel(userID int) *model.Generator {
	return &model.Generator{
		ID:          g.ID,
		UserID:      userID,
		Name:        g.Name,
		Description: g.Description,
		Interval:    time.Duration(g.IntervalDays) * 24 * time.Hour,
		Params:      g.Params.ToModel(),
	}
}

File: ui/src/lib/types/generator.ts
import z from "zod";
import { JSONBody } from "./general";
import { API } from "./api";
import { convertTracks, Track } from "./track";

export enum GeneratorPreset {
  Top = "top",
  OldTop = "old_top"
}
export const generatorPresetString: Record<GeneratorPreset, string> = {
  [GeneratorPreset.Top]: "Top",
  [GeneratorPreset.OldTop]: "Old Top",
}

export interface GeneratorWindow {
  start: Date;
  end: Date;
  minPlays: number;
  burstIntervalDays: number;
}

export const convertGeneratorWindow = (g: API.GeneratorWindow): GeneratorWindow => {
  return {
    start: new Date(g.start),
    end: new Date(g.end),
    minPlays: g.min_plays,
    burstIntervalDays: g.burst_interval_days,
  }
}

export interface GeneratorParams {
  trackAmount: number;
  excludedPlaylistIds: number[];
  excludedTrackIds: number[];
  preset: GeneratorPreset;
  paramsTop?: {
    window: GeneratorWindow;
  };
  paramsOldTop?: {
    peakWindow: GeneratorWindow;
    recentWindow: GeneratorWindow;
  };
}

export const convertGeneratorParams = (g: Pick<API.Generator, "params">): GeneratorParams => {
  return {
    trackAmount: g.params.track_amount,
    excludedPlaylistIds: g.params.excluded_playlist_ids ?? [],
    excludedTrackIds: g.params.excluded_track_ids ?? [],
    preset: g.params.preset as GeneratorPreset,
    paramsTop: g.params.params_top ? {
      window: convertGeneratorWindow(g.params.params_top.window)
    } : undefined,
    paramsOldTop: g.params.params_old_top ? {
      peakWindow: convertGeneratorWindow(g.params.params_old_top.peak_window),
      recentWindow: convertGeneratorWindow(g.params.params_old_top.recent_window),
    } : undefined,
  }
}

export interface Generator {
  id: number;
  name: string;
  description?: string;
  playlistId?: number;
  intervalDays: number;
  spotifyOutdated: boolean;
  params: GeneratorParams;
  tracks: Track[];
  lastUpdate?: Date;
}

export const convertGenerator = (g: API.Generator): Generator => {
  return {
    id: g.id,
    name: g.name,
    description: g.description,
    playlistId: g.playlist_id,
    intervalDays: g.interval_days,
    spotifyOutdated: g.spotify_outdated,
    params: convertGeneratorParams(g),
    tracks: convertTracks(g.tracks),
    lastUpdate: g.last_update ? new Date(g.last_update) : undefined,
  }
}

export const convertGenerators = (g: API.Generator[]): Generator[] => {
  return g.map(convertGenerator)
}

export const convertGeneratorSchema = (g: Generator): GeneratorSchema => {
  return {
    id: g.id,
    name: g.name,
    description: g.description,
    createPlaylist: (g.playlistId ?? 0) !== 0,
    intervalDays: g.intervalDays,
    params: { ...g.params },
  }
}

export const generatorWindowSchema = z.object({
  start: z.date(),
  end: z.date(),
  minPlays: z.number().positive(),
  burstIntervalDays: z.number().positive(),
})
export type GeneratorWindowSchema = z.infer<typeof generatorWindowSchema> & JSONBody;

export const generatorParamsSchema = z.object({
  trackAmount: z.number().positive(),
  excludedPlaylistIds: z.array(z.number().positive()),
  excludedTrackIds: z.array(z.number().positive()),
  preset: z.enum(GeneratorPreset),
  paramsTop: z.object({
    window: generatorWindowSchema
  }).partial().optional(),
  paramsOldTop: z.object({
    peakWindow: generatorWindowSchema,
    recentWindow: generatorWindowSchema
  }).partial().optional(),
})
export type GeneratorParamsSchema = z.infer<typeof generatorParamsSchema> & JSONBody;

export const generatorSchema = z.object({
  id: z.number().positive().optional(),
  name: z.string().nonempty(),
  description: z.string().optional(),
  createPlaylist: z.boolean(),
  intervalDays: z.number().nonnegative(),
  params: generatorParamsSchema.partial().optional(),
})
export type GeneratorSchema = z.infer<typeof generatorSchema> & JSONBody;
